File: go/ir/event.go
package ir

import "encoding/json"

// Event is a streaming event (spec/01 s5.1). Sealed: exactly six variants in
// v1. Type definitions only in this milestone; streaming logic arrives in
// M6/M7.
type Event interface {
	isEvent()
}

// MessageStart opens the stream.
type MessageStart struct {
	ID    string
	Model string
}

// ContentBlockStart opens a content block at Index.
type ContentBlockStart struct {
	Index int
	Block Block
}

// ContentBlockDelta carries a delta for the currently open block.
type ContentBlockDelta struct {
	Index int
	Delta Delta
}

// ContentBlockStop closes the currently open block.
type ContentBlockStop struct {
	Index int
}

// MessageDelta carries the final stop reason and usage, immediately before
// MessageDone (INV-5).
type MessageDelta struct {
	StopReason   StopReason
	StopSequence string // same conditional rule as Response.StopSequence
	Usage        Usage
}

// MessageDone terminates the stream.
type MessageDone struct{}

func (MessageStart) isEvent()      {}
func (ContentBlockStart) isEvent() {}
func (ContentBlockDelta) isEvent() {}
func (ContentBlockStop) isEvent()  {}
func (MessageDelta) isEvent()      {}
func (MessageDone) isEvent()       {}

// Delta is the payload of a ContentBlockDelta (spec/01 s5.2). Sealed: exactly
// two variants in v1.
type Delta interface {
	isDelta()
}

// TextDelta is a text fragment.
type TextDelta struct {
	Text string
}

// InputJSONDelta is a fragment of the tool-argument string. PartialJSON is
// raw JSON text held as the exact source JSON string token (INV-1): it is
// never parsed or re-serialized, and concatenation of all fragments of a
// block is the block's input.
type InputJSONDelta struct {
	PartialJSON json.RawMessage
}

func (TextDelta) isDelta()      {}
func (InputJSONDelta) isDelta() {}

// EventStream is the JSON document form of a streamed response (spec/01
// s5.3): specVersion plus the totally ordered event sequence.
type EventStream struct {
	Events []Event
}

File: rust/crates/oxa-ir/src/event.rs
//! Streaming event types (spec/01 §5).

use serde::{Deserialize, Serialize};

use crate::request::Block;
use crate::response::{StopReason, Usage};

/// One streaming event (spec/01 §5.1). Sealed; exactly six variants in v1,
/// discriminated on the JSON `type` property. The event sequence obeys the
/// INV-5 grammar and INV-6 index discipline, enforced by [`crate::checker`].
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum Event {
    #[serde(rename = "message_start")]
    MessageStart {
        #[serde(rename = "id")]
        id: String,
        #[serde(rename = "model")]
        model: String,
    },
    #[serde(rename = "content_block_start")]
    ContentBlockStart {
        #[serde(rename = "index")]
        index: i64,
        #[serde(rename = "block")]
        block: Block,
    },
    #[serde(rename = "content_block_delta")]
    ContentBlockDelta {
        #[serde(rename = "index")]
        index: i64,
        #[serde(rename = "delta")]
        delta: Delta,
    },
    #[serde(rename = "content_block_stop")]
    ContentBlockStop {
        #[serde(rename = "index")]
        index: i64,
    },
    #[serde(rename = "message_delta")]
    MessageDelta {
        #[serde(rename = "stop_reason")]
        stop_reason: StopReason,
        #[serde(
            rename = "stop_sequence",
            default,
            skip_serializing_if = "Option::is_none"
        )]
        stop_sequence: Option<String>,
        #[serde(rename = "usage")]
        usage: Usage,
    },
    #[serde(rename = "message_done")]
    MessageDone {},
}

/// The delta payload of a `content_block_delta` event (spec/01 §5.2).
/// `InputJSONDelta::partial_json` is opaque raw JSON text (INV-1).
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum Delta {
    #[serde(rename = "text_delta")]
    TextDelta {
        #[serde(rename = "text")]
        text: String,
    },
    #[serde(rename = "input_json_delta")]
    InputJsonDelta {
        #[serde(rename = "partial_json")]
        partial_json: String,
    },
}

/// The JSON document form of a streamed response (spec/01 §5.3).
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct EventStream {
    #[serde(rename = "events")]
    pub events: Vec<Event>,
}
